Feature: Floating turbine Cp/Ct/Tilt support

bayc opened the pull request:
**Feature or improvement description**
This PR adds support for floating turbines in the form of Cp/Ct curve corrections for tilt of the floating platform based on wind speed.

**Related issue, if one exists**
#515
#517 
#526 
#527 

**Impacted areas of the software**
A lot of the simulation code, focusing mostly in `turbine.py` and `farm.py`. Updates to `floris_interface.py` as well as the pre-defined turbines.

**Additional supporting information**
Specific tasks involved in this PR:

- [x] Add reference tilt to inputs and `turbine.py`
- [x] Add tilt/wind speed table to inputs and `turbine.py`
- [x] Build out tilt values in `farm.py`
- [x] Update solvers to support tilt values/corrections
- [x] Re-scope power function in `turbine.py`
- [x] Add tilt angle support to `floris_interface.py`
- [x] Update all predefined turbine definitions to include reference tilt
- [x] Update unit tests
- [x] Update regression tests

**Test results, if applicable**
Currently all tests pass. Unit tests were added for the new/rescoped functions in `turbine.py`, while other unit tests were updated appropriately for tilt inputs. Note the regression test baseline values did not change. This was intentional to show that the tilt correction code does not cause a change in calculated values when the tilt of the turbine is equal to the reference tilt of that turbine.

paulf81 commented:
Yes @bayc I think you've captured it all nicely, this looks good!

rafmudaf commented on .gitignore:
```diff
@@ -32,3 +32,9 @@ examples/SLSQP.out
 # Log files
 *.log
+
+# User-specific files
```
Are these relevant to keep or left over from local changes?

bayc replied:
Local changes, good catch! They are removed now.

rafmudaf commented on examples/inputs_floating/gch_fixed.yaml:
```diff
@@ -0,0 +1,226 @@
+
```
The gch.yaml input file also serves as the reference for the documentation. It might be worthwhile to remove all of the comments here for readability.
Same comment for the group of input files added here.

rafmudaf commented on floris/simulation/turbine.py:
```diff
@@ -319,6 +489,38 @@ def __attrs_post_init__(self) -> None:
+
+        # Remove any duplicate wind speed entries
+        _, duplicate_filter = np.unique(self.wind_speeds, return_index=True)
+        object.__setattr__(self, "tilt", self.tilt[duplicate_filter])
```
Just curious about these two lines. What's different between this and the following;
```python
self.tilt = self.tilt[duplicate_filter]
```

bayc replied:
After talking with @RHammond2, my understanding is that this method was used when these tables were initially immutable (I modeled this after the pattern in PowerThrustTable). But sounds like the change was made along the way to have PowerThrustTable to be mutable, so I've updated to your suggestion in both tables.

rafmudaf commented on tests/turbine_unit_test.py:
```diff
@@ -373,4 +576,8 @@ def test_asdict(sample_inputs_fixture: SampleInputs):
     new_turb = Turbine.from_dict(dict1)
     dict2 = new_turb.as_dict()
 
+    print(dict1)
```
Still needed?

bayc replied:
Nope. Removed!

rafmudaf reviewed (changes requested):
I see that this pull request is still a draft, but I looked through the changes and wanted to mentioned a few things. Also, please update the examples listing [here](https://github.com/NREL/floris/blob/develop/docs/examples.md) describing the new example.

bayc commented:
> I see that this pull request is still a draft, but I looked through the changes and wanted to mentioned a few things. Also, please update the examples listing [here](https://github.com/NREL/floris/blob/develop/docs/examples.md) describing the new example.

Updated and added!

rafmudaf commented on floris/simulation/farm.py:
```diff
@@ -25,6 +25,7 @@
     State,
     Turbine,
 )
+from floris.simulation.turbine import _compute_tilt_angles_for_floating_turbines
```
Since `_compute_tilt_angles_for_floating_turbines` is imported outside of turbine, it should not have a prefixed underscore, or there's a reason it is meant to be private it should not be imported directly here.

bayc replied:
Updated to be a public method.

rafmudaf replied:
I think you may have missed that change in your commit, I still see it with the underscore in the pull request

bayc replied:
Had it locally, didn't push it up. Should be there now.

rafmudaf commented on floris/simulation/farm.py:
```diff
@@ -131,14 +135,27 @@ def construct_rotor_diameters(self):
 
-    def construc_turbine_ref_density_cp_cts(self):
+    def construct_turbine_pTs(self):
+        self.pTs = np.array([turb['pT'] for turb in self.turbine_definitions])
```
This should be listed in the attributes for this class at the top.

rafmudaf commented on floris/simulation/farm.py:
```diff
@@ -131,14 +135,27 @@ def construct_rotor_diameters(self):
 
-    def construc_turbine_pPs(self):
+    def construct_turbine_pPs(self):
         self.pPs = np.array([turb['pP'] for turb in self.turbine_definitions])
```
This should be listed in the attributes for this class at the top.

rafmudaf commented on floris/simulation/farm.py:
```diff
@@ -131,14 +135,27 @@ def construct_rotor_diameters(self):
         ])
 
+    def construct_turbine_ref_tilt_cp_cts(self):
+        self.ref_tilt_cp_cts = np.array(
```
This should be listed in the attributes for this class at the top.

rafmudaf commented on floris/simulation/farm.py:
```diff
@@ -131,14 +135,27 @@ def construct_rotor_diameters(self):
+        )
+
+    def construct_turbine_correct_cp_ct_for_tilt(self):
+        self.correct_cp_ct_for_tilt = np.array(
```
This should be listed in the attributes for this class at the top.

rafmudaf commented on floris/simulation/farm.py:
```diff
@@ -148,6 +165,9 @@ def construct_turbine_fCts(self):
         self.turbine_fCps = [(turb.turbine_type, turb.fCp_interp) for turb in self.turbine_map]
 
+    def construct_turbine_fTilts(self):
+        self.turbine_fTilts = [(turb.turbine_type, turb.fTilt_interp) for turb in self.turbine_map]
```
This should be listed in the attributes for this class at the top.

rafmudaf commented on floris/simulation/farm.py:
```diff
@@ -180,11 +200,36 @@ def expand_farm_properties(
             sorted_coord_indices,
             axis=2
         )
+        self.ref_density_cp_cts_sorted = np.take_along_axis(
```
Need a similar method for fTilts?

bayc replied:
Using the pattern from Ct/power interpolants, an array of just the different interpolators is kept (one for each kind of turbine) and the turbine type is checked to know which interpolant (fTilt) to use, so don't need to sort them.

rafmudaf commented on floris/tools/floris_interface.py:
```diff
@@ -630,13 +658,18 @@ def get_turbine_ais(self) -> NDArrayFloat:
 
     @property
     def turbine_average_velocities(self) -> NDArrayFloat:
+        # TODO: Should we have a `turbine_effective_velocities` function?
```
Will this be used in analysis? If so, it would be good to have.

bayc replied:
@rafmudaf Unsure if it would be used or not. It isn't currently, that I know of. The `turbine_average_velocities` are used in a couple of examples, but it doesn't include the effect of yaw or tilt (which would be included in the rotor effective velocity). @paulf81 What do you think?

rafmudaf commented on floris/simulation/turbine.py:
```diff
@@ -125,34 +257,21 @@ def power(
     # That said, it adds overhead to convert the floats to 1-D arrays, so I don't
```
I think deleting the lines below this means this comment block is no longer relevant and can be deleted

rafmudaf commented on floris/simulation/turbine.py:
```diff
@@ -162,7 +281,11 @@ def power(
 def Ct(
     velocities: NDArrayFloat,
     yaw_angle: NDArrayFloat,
+    tilt_angle: NDArrayFloat,
```
Need to add these arguments to the docstring below:
tilt_angle
ref_tilt_cp_ct
tilt_interp
correct_cp_ct_for_tilt

rafmudaf commented on floris/simulation/turbine.py:
```diff
@@ -210,14 +350,18 @@ def Ct(
 def axial_induction(
     velocities: NDArrayFloat,  # (wind directions, wind speeds, turbines, grid, grid)
     yaw_angle: NDArrayFloat,  # (wind directions, wind speeds, turbines)
+    tilt_angle: NDArrayFloat,  # (wind directions, wind speeds, turbines)
```
Need to add these arguments to the docstring

rafmudaf commented on floris/simulation/turbine.py:
```diff
@@ -359,17 +561,22 @@ class Turbine(BaseClass):
     power_thrust_table: PowerThrustTable = field(converter=PowerThrustTable.from_dict)
-
-
+    # floating_tilt_table: NDArrayFloat = field(
```
It looks like this is initialized below so this commented code is probably out of date

bayc commented on floris/tools/floris_interface.py:
```diff
@@ -642,6 +662,23 @@ def get_turbine_ais(self) -> NDArrayFloat:
         return average_velocity(velocities=self.floris.flow_field.u)
 
+    @property
+    def turbine_effective_velocities(self) -> NDArrayFloat:
```
@rafmudaf Added `turbine_effective_velocities` at the `floris_interface` level. Should this be an `@property`? Modelled after `turbine_average_velocities`.

paulf81 commented:
The latest commit [adding turbine_effective_velocities to floris_interface](https://github.com/NREL/floris/pull/518/commits/138f56a74367da0325aed6ed5f6470f0668765a3) I think is causing example 24 to break,

bayc commented:
> The latest commit [adding turbine_effective_velocities to floris_interface](https://github.com/NREL/floris/pull/518/commits/138f56a74367da0325aed6ed5f6470f0668765a3) I think is causing example 24 to break,

Thanks @paulf81, I've updated so that all the examples work now.

paulf81 commented:
Hi @bayc , I was thinking I was going to switch the example to use an external library folder, instead of the !include syntax to reference the floating turbine model, but I think this path would have to be specified so then it seems better actually to leave these as is for now, and maybe we can include syntax on using the external library folder to documentation rather than examples, what do you think @rafmudaf ?
